### PR TITLE
dynamic inheritance for packages that change build system

### PR DESCRIPTION
**Pull request by becker33**
Began with @gonsie changing mpifileutils to use cmake.

Implemented DynamicInheritancePackage build system that can dynamically change its phases and attributes to mirror the other build system classes. The `classes` field is a dictionary of build system classes and anonymous specs describing when they are used.

Package classes are responsible for shadowing any dependencies and variants from the build system classes they intend to inherit from.

mpifileutils provides an example with a package changing from autotools to cmake, including handling the `build_type` variant from `CMakePackage` through the flag_handler method for the earlier versions that use autotools.

@tgamblin I think you will find this interesting.

### CODE REVIEW

File: lib/spack/spack/build_systems/dynamic_inheritance.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+                make_package_build_system(self, cls)
+
+        # Remove phases not used by this build system
+        # keep phases we have methods for and build system phases
+        for phase in self.phases:
+            if phase not in self.__dict__ and phase not in self.cls.phases:
+                setattr(self, '_InstallPhase_%s' % phase, None)
+
+    def unset_build_system(self, spec, prefix):
+        # This probably doesn't overwrite InstallPhse objects properly.
```
Comment (healther) [inline lib/spack/spack/build_systems/dynamic_inheritance.py]: InstallPhase

File: lib/spack/spack/build_systems/dynamic_inheritance.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+def steal_attribute(pkg, attribute):
+    """
+    Attaches attribute to pkg, while maintaining type information.
+
+    Methods are turned to functions and attached via MethodType
+    Properties are captured using fget and attached to the package class
+    All others are attached directly.
+
+    This will need to be revisited if/when a build system adds a multimethod
```
Comment (healther) [inline lib/spack/spack/build_systems/dynamic_inheritance.py]: This assumption should probably be guarded by a unit test...

File: var/spack/repos/builtin/packages/mpifileutils/package.py [hunk 3]
```diff
@@ -3,10 +3,8 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
-from spack import *
```
Comment (healther) [inline var/spack/repos/builtin/packages/mpifileutils/package.py]: You probably don't want to delete this line

File: var/spack/repos/builtin/packages/mpifileutils/package.py [hunk 4]
```diff
@@ -39,18 +42,74 @@ class Mpifileutils(AutotoolsPackage):
 
     depends_on('libarchive')
 
+    depends_on('cmake@3.1', when='@0.9:', type='build')
+
+    variant('build_type', default='RelWithDebInfo',
```
Comment (healther) [inline var/spack/repos/builtin/packages/mpifileutils/package.py]: Doesn't this variant come from CMakePackage? Why does it need to be created here again?

**Comment by alalazo** [comment]
**TL;DR** I think this PR addresses a need that is present in more than a few packages and has been overlooked so far, but I also have a few strong concerns on the current implementation where all the attributes of different build-systems are merged in a single class. 

This can become quite messy and doesn't keep the information pertaining to different build-systems separated from each other. To get this feature right we might need some support from the concretizer and possibly an extension of the spec syntax. 

I'll share below a few random thoughts that I noted down in the past on this issue. Hope they can be useful to foster some discussion.

---

Packages that can be built with different build systems fall into two categories:

1. Software that changed build-system at a particular point in time (e.g. `mpifileutils`)
2. Software that for a non-null time window could be built with two different build systems (e.g. `hdf5` or `fftw`)

In case number 1. we have a 1:1 correspondence from a version number to a build-system. Case number 2. is more problematic as for a given version there can be multiple build-systems. The latter case might not be resolved with a "pick your preferred build-system" approach, as the artifacts of the build might be different for the same version of the software. 

For all the examples that I could find right now, the time period in which a package could be built with a given build-system can be mapped to a version range. Nothing prevents though to identify this set with further constraints (like the `os` or anything - think of a package using one build system on Linux and one on Windows). Not sure though if generalizing this aspect is only of theoretical interest or if there are actual use cases.

If we model packages supporting multiple build-systems under the assumption that for a given version number we have a corresponding build-system (restrictive with respect to the use cases found so far) we can employ the current spec syntax. If we go instead with overlapping build-systems that the user can choose, the spec syntax needs to be extended (e.g. `fftw` can currently be built with CMake and Autotools, the user needs to be able to specify which one he wants to use).

Different build-systems might mean different variants or even different dependencies for the package. Nonetheless it would be good to keep the information well factored and avoid mixing together variants or attributes that come from different places. A tentative package API could be obtained with decorators (the pseudo-code below is only for illustrative purposes):
```python
class FFTW(MultipleBuildSystems):
    """FFTW is ..."""

    # All the common directives and attributes


@FFTW.when('@:3.1')
class _FFTW(AutotoolsPackage):

    # All the directives and attributes related to the autotools build-system only
    # Here I need to be able to refer to variants in the FFTW main class


@FFTW.when('@3.0:')
class _FFTW(CMakePackage):

    # All the directives and attributes related to the CMake build-system only
    # Here I need to be able to refer to variants in the FFTW main class

```

The FFTW class should behave, during concretization, as a finite state machine with 2 states:

- build-system-undefined 
- build-system-selected

When in the `build-system-undefined` state it should accept any constraint imposed by the concretizer, as the class doesn't know yet its full set of dependencies or variants etc. As soon as the build-system can be selected the constraints that have been imposed so far need to be checked and any inconsistency needs to be reported. This will probably need some support from the concretizer. 

At install time the underlying spec is already concretized, and the behavior of the class is well-defined, so that should not pose any problem.

File: var/spack/repos/builtin/packages/mpifileutils/package.py (hunk 3 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/mpifileutils/package.py]: I actually did. That line has been unnecessary for a while, since #8090. But @tgamblin worked out the fix so it would be backwards compatible so we wouldn't have to update all the packages.

File: var/spack/repos/builtin/packages/mpifileutils/package.py (hunk 4 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/mpifileutils/package.py]: Packages inheriting from `DynamicInheritancePackage` will have to shadow the directives from the parent class, since they need access to those before they get a chance to set their inheritance. That's variants, dependencies, and conflicts. If anyone has a brilliant idea let me know, but as of now I don't know of any way to do it other than shadowing those.

**Comment by AndrewGaspar** [comment]
Just want to leave some input that this would be useful for a few packages that transitioned from merely header-only to support a CMake style package system (supporting `find_package`).

**Comment by alalazo** [comment]
Just want to mention a simpler approach for the same issue in #15420 meant for comparison with other proposals.

**Comment by alalazo** [comment]
@becker33 I'm closing this PR and I added a link to it in https://github.com/spack/seps/pull/3
